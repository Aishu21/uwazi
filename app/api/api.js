import bodyParser from 'body-parser'
import db_config from './config/database.js'

export default app => {

  //set db to use
  db_config.db_url = db_config.production;

  //common middlewares
  app.use(bodyParser.json());

  //module routes
  require('./auth/routes.js')(app);

}
