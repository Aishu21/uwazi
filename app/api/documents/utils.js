export function updateMetadataNames(documents, nameMatches) {
  documents.forEach((doc) => {
    let oldMetadata = Object.assign({}, doc.metadata);
    Object.keys(doc.metadata).forEach((propertyName) => {
      if (nameMatches[propertyName]) {
        delete doc.metadata[propertyName];
        doc.metadata[nameMatches[propertyName]] = oldMetadata[propertyName];
      }
    });
  });

  return documents;
}
