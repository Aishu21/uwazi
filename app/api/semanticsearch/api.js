import fetch from 'isomorphic-fetch';
import SEMANTIC_SEARCH_URL from 'api/config/semanticsearch';

const semanticSearchAPI = {
  async processDocument(args) {
    const res = await fetch(`${SEMANTIC_SEARCH_URL}/semanticSearch/searchOneDoc`, {
      method: 'POST',
      body: JSON.stringify(args),
      headers: {
        'Content-Type': 'application/json'
      }
    });
    const body = await res.json();
    return body;
  }
};

export default semanticSearchAPI;
