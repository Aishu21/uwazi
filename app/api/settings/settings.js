import {db_url as dbUrl} from 'api/config/database';
import request from 'shared/JSONRequest';

export default {
  get() {
    return request.get(`${dbUrl}/_design/settings/_view/all`)
    .then((result) => {
      return result.json.rows[0].value;
    });
  },

  save(settings) {
    return request.post(dbUrl, settings)
    .then((result) => {
      return result.json;
    });
  }
};
