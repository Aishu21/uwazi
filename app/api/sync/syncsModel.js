import mongoose from 'mongoose';

const syncSchema = new mongoose.Schema({
  lastSync: Number,
});

export default mongoose.model('syncs', syncSchema);

