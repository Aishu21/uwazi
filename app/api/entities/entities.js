import {db_url as dbURL} from 'api/config/database';
import request from 'shared/JSONRequest';
import {updateMetadataNames, deleteMetadataProperties} from 'api/entities/utils';
import date from 'api/utils/date.js';
import sanitizeResponse from '../utils/sanitizeResponse';
import references from '../references/references.js';
import settings from '../settings';
import ID from 'shared/uniqueId';

export default {
  save(doc, {user, language}) {
    doc.type = 'entity';
    if (!doc._id) {
      doc.user = user;
      doc.creationDate = date.currentUTC();
    }

    const sharedId = doc.sharedId || ID();
    return settings.get()
    .then(({languages}) => {
      if (doc._id) {
        return request.post(dbURL + '/_design/entities/_update/partialUpdate/' + doc._id, doc);
      }

      const docs = languages.map((lang) => {
        let langDoc = Object.assign({}, doc);
        langDoc.language = lang.key;
        langDoc.sharedId = sharedId;
        return langDoc;
      });

      return request.post(`${dbURL}/_bulk_docs`, {docs});
    })
    .then(() => request.get(`${dbURL}/_design/entities/_view/by_language`, {key: [sharedId, language]}))
    .then(response => {
      return Promise.all([response, references.saveEntityBasedReferences(response.json.rows[0].value)]);
    })
    .then(([response]) => response.json.rows[0].value);
  },

  countByTemplate(templateId) {
    return request.get(`${dbURL}/_design/entities/_view/count_by_template?group_level=1&key="${templateId}"`)
    .then((response) => {
      if (!response.json.rows.length) {
        return 0;
      }
      return response.json.rows[0].value;
    });
  },

  getByTemplate(templateId) {
    return request.get(`${dbURL}/_design/entities/_view/by_template?&key="${templateId}"`)
    .then((response) => {
      return sanitizeResponse(response.json);
    });
  },

  updateMetadataProperties(templateId, nameMatches, deleteProperties) {
    return request.get(`${dbURL}/_design/entities/_view/metadata_by_template?key="${templateId}"`)
    .then((response) => {
      let entities = response.json.rows.map((r) => r.value);
      entities = updateMetadataNames(entities, nameMatches);
      entities = deleteMetadataProperties(entities, deleteProperties);

      let updates = [];
      entities.forEach((entity) => {
        let url = `${dbURL}/_design/entities/_update/partialUpdate/${entity._id}`;
        updates.push(request.post(url, entity));
      });

      return Promise.all(updates);
    });
  },

  delete(id) {
    let docsToDelete = [];
    return request.get(`${dbURL}/${id}`)
    .then((response) => {
      docsToDelete.push({_id: response.json._id, _rev: response.json._rev});
      return request.get(`${dbURL}/_design/references/_view/by_source?key="${id}"`);
    })
    .then((response) => {
      sanitizeResponse(response.json);
      docsToDelete = docsToDelete.concat(response.json.rows);
      return request.get(`${dbURL}/_design/references/_view/by_target?key="${id}"`);
    })
    .then((response) => {
      sanitizeResponse(response.json);
      docsToDelete = docsToDelete.concat(response.json.rows);
      docsToDelete.map((doc) => doc._deleted = true);
      return request.post(`${dbURL}/_bulk_docs`, {docs: docsToDelete});
    })
    .then((response) => {
      return response.json;
    });
  }
};
