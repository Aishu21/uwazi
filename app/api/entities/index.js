import entities from './entities.js';
import model from './entitiesModel.js';
import endpointSchema from './endpointSchema';
export default entities;

export {
  model,
  endpointSchema
};
