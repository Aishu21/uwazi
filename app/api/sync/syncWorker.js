import 'api/entities';

import urljoin from 'url-join';

import { models } from 'api/odm';
import { prettifyError } from 'api/utils/handleError';
import { model as updateLog } from 'api/updatelogs';
import errorLog from 'api/log/errorLog';
import request from 'shared/JSONRequest';
import settings from 'api/settings';

import syncsModel from './syncsModel';

const oneSecond = 1000;

const timeout = async interval => new Promise((resolve) => {
  setTimeout(resolve, interval);
});

const syncData = async (url, action, change, data) => {
  await request[action](urljoin(url, 'api/sync'), { namespace: change.namespace, data });
  return syncsModel.updateMany({}, { $set: { lastSync: change.timestamp } });
};

export default {
  stopped: false,

  async syncronize({ url, config }) {
    const [{ lastSync }] = await syncsModel.find();
    const lastChanges = await updateLog.find({
      timestamp: {
        $gte: lastSync - oneSecond
      },
      namespace: {
        $nin: ['migrations']
      }
    }, null, {
      sort: {
        timestamp: 1
      },
      lean: true
    });

    // there is always one ??
    // console.log(lastChanges[0]);

    await lastChanges.reduce(async (prev, change) => {
      await prev;
      if (change.deleted) {
        return syncData(url, 'delete', change, { _id: change.mongoId });
      }

      const templatesConfig = config.templates || {};

      if (change.namespace === 'templates' && !templatesConfig[change.mongoId.toString()]) {
        return Promise.resolve();
      }

      const data = await models[change.namespace].getById(change.mongoId);

      if (change.namespace === 'entities' && !Object.keys(templatesConfig).includes(data.template.toString())) {
        return Promise.resolve();
      }


      if (change.namespace === 'templates' && data.properties) {
        data.properties = data.properties.filter(property => templatesConfig[data._id.toString()].includes(property._id.toString()));
      }

      if (change.namespace === 'entities' && data.metadata) {
        const templateData = await models.templates.getById(data.template);

        const validPropertyNames = templateData.properties.reduce((memo, property) => {
          if (templatesConfig[templateData._id.toString()].includes(property._id.toString())) {
            memo.push(property.name);
          }
          return memo;
        }, []);

        data.metadata = Object.keys(data.metadata).reduce((_memo, propertyName) => {
          const memo = _memo;
          if (validPropertyNames.includes(propertyName)) {
            memo[propertyName] = data.metadata[propertyName];
          }
          return memo;
        }, {});
      }

      return syncData(url, 'post', change, data);
    }, Promise.resolve());
  },

  async intervalSync(url, interval = 5000) {
    if (this.stopped) {
      return;
    }
    try {
      await this.syncronize(url);
    } catch (e) {
      if (e.status === 401) {
        await this.login(url, 'admin', 'admin');
      } else {
        errorLog.error(prettifyError(e).prettyMessage);
      }
    }
    await timeout(interval);
    await this.intervalSync(url, interval);
  },

  async login(url, username, password) {
    try {
      const response = await request.post(urljoin(url, 'api/login'), { username, password });
      request.cookie(response.cookie);
    } catch (e) {
      errorLog.error(prettifyError(e).prettyMessage);
    }
  },

  async start(interval) {
    const { sync } = await settings.get();
    if (sync && sync.active) {
      this.intervalSync(sync.url, interval);
    }
  },

  stop() {
    this.stopped = true;
  }
};
