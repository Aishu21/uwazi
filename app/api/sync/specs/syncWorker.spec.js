import 'api/entities';
import errorLog from 'api/log/errorLog';
import 'api/relationships';
import backend from 'fetch-mock';

import { catchErrors } from 'api/utils/jasmineHelpers';
import db from 'api/utils/testing_db';
import request from 'shared/JSONRequest';
import settings from 'api/settings';
import settingsModel from 'api/settings/settingsModel';

import fixtures, {
  newDoc1,
  newDoc2,
  newDoc3,
  newDoc4,
  template1,
  template1Property1,
  template1Property2,
  template1Property3,
  template2,
  template3,
} from './fixtures';
import syncWorker from '../syncWorker';
import syncsModel from '../syncsModel';

describe('syncWorker', () => {
  beforeEach((done) => {
    spyOn(errorLog, 'error');
    db.clearAllAndLoad(fixtures).then(done).catch(catchErrors(done));
    syncWorker.stopped = false;
  });

  afterAll((done) => {
    db.disconnect().then(done);
  });

  const syncAllTemplates = async () => syncWorker.syncronize({
    url: 'url',
    config: {
      templates: {
        [template1.toString()]: [],
        [template2.toString()]: [],
        [template3.toString()]: [],
      }
    }
  });

  describe('syncronize', () => {
    it('should lazy create lastSync entry if not exists', async () => {
      spyOn(request, 'post').and.callFake(() => Promise.reject());
      spyOn(request, 'delete').and.callFake(() => Promise.reject());

      await syncsModel.remove({});

      try {
        await syncWorker.syncronize({ url: 'url', config: {} });
      } catch (e) {
        const [{ lastSync }] = await syncsModel.find();
        expect(lastSync).toBe(0);
      }
    });

    describe('templates', () => {
      it('should only sync white listed templates and properties', async () => {
        spyOn(request, 'post').and.returnValue(Promise.resolve());
        spyOn(request, 'delete').and.returnValue(Promise.resolve());

        await syncWorker.syncronize({
          url: 'url',
          config: {
            templates: {
              [template1.toString()]: [template1Property1.toString(), template1Property3.toString()],
              [template2.toString()]: [],
            }
          }
        });

        const templateCallsOnly = request.post.calls.allArgs().filter(args => args[1].namespace === 'templates');
        const template1Call = templateCallsOnly.find(call => call[1].data._id.toString() === template1.toString());
        const template2Call = templateCallsOnly.find(call => call[1].data._id.toString() === template2.toString());

        expect(templateCallsOnly.length).toBe(2);

        expect(template1Call).toEqual(['url/api/sync', {
          namespace: 'templates',
          data: {
            _id: template1,
            properties: [
              expect.objectContaining({ _id: template1Property1 }),
              expect.objectContaining({ _id: template1Property3 })
            ]
          }
        }]);
        expect(template2Call).toEqual(['url/api/sync', {
          namespace: 'templates',
          data: { _id: template2 }
        }]);
      });
    });

    describe('entities', () => {
      it('should only sync entities belonging to a white listed template and properties', async () => {
        spyOn(request, 'post').and.returnValue(Promise.resolve());
        spyOn(request, 'delete').and.returnValue(Promise.resolve());

        await syncWorker.syncronize({
          url: 'url',
          config: {
            templates: {
              [template1.toString()]: [template1Property2.toString(), template1Property3.toString()],
              [template2.toString()]: [],
            }
          }
        });

        const entitiesCallsOnly = request.post.calls.allArgs().filter(args => args[1].namespace === 'entities');
        const entity1Call = entitiesCallsOnly.find(call => call[1].data._id.toString() === newDoc1.toString());
        const entity2Call = entitiesCallsOnly.find(call => call[1].data._id.toString() === newDoc2.toString());

        expect(entity1Call).toEqual(['url/api/sync', {
          namespace: 'entities',
          data: expect.objectContaining({
            _id: newDoc1,
            metadata: {
              t1Property2: 'sync property 2',
              t1Property3: 'sync property 3',
            }
          })
        }]);

        expect(entity2Call).toEqual(['url/api/sync', {
          namespace: 'entities',
          data: expect.objectContaining({
            _id: newDoc2,
            metadata: { t1Property2: 'another doc property 2' },
          })
        }]);

        expect(request.post).not.toHaveBeenCalledWith('url/api/sync', {
          namespace: 'entities',
          data: expect.objectContaining({ title: 'not to sync' })
        });
      });
    });

    it('should process the log records newer than the current sync time (minus 1 sec)', async () => {
      spyOn(request, 'post').and.returnValue(Promise.resolve());
      spyOn(request, 'delete').and.returnValue(Promise.resolve());

      await syncAllTemplates();

      expect(request.post.calls.count()).toBe(8);
      expect(request.delete.calls.count()).toBe(1);
    });

    it('should update lastSync timestamp with the last change', async () => {
      spyOn(request, 'post').and.returnValue(Promise.resolve());
      spyOn(request, 'delete').and.returnValue(Promise.resolve());

      await syncAllTemplates();
      const [{ lastSync }] = await syncsModel.find();
      expect(lastSync).toBe(22000);
    });

    it('should update lastSync on each operation', async () => {
      spyOn(request, 'post').and.callFake((url, body) =>
        body.data._id.equals(newDoc3) ? Promise.reject(new Error('post failed')) : Promise.resolve()
      );
      spyOn(request, 'delete').and.callFake((url, body) =>
        body.data._id.equals(newDoc4) ? Promise.reject(new Error('delete failed')) : Promise.resolve()
      );

      try {
        await syncAllTemplates();
      } catch (e) {
        const [{ lastSync }] = await syncsModel.find();
        expect(lastSync).toBe(12000);
      }
    });
  });

  describe('intervalSync', () => {
    it('should syncronize every x seconds', async () => {
      let syncCalls = 0;
      spyOn(syncWorker, 'syncronize').and.callFake(() => {
        if (syncCalls === 2) {
          syncWorker.stop();
        }
        syncCalls += 1;
        return Promise.resolve();
      });

      const interval = 0;
      await syncWorker.intervalSync({ url: 'url' }, interval);
      expect(syncWorker.syncronize).toHaveBeenCalledWith({ url: 'url' });
      expect(syncWorker.syncronize).toHaveBeenCalledTimes(3);
    });

    it('should retry when syncronize returns a request error', async () => {
      let syncCalls = 0;
      spyOn(syncWorker, 'syncronize').and.callFake(() => {
        if (syncCalls === 2) {
          syncWorker.stop();
        }
        syncCalls += 1;
        return Promise.reject({ status: 500, message: 'error' }); // eslint-disable-line prefer-promise-reject-errors
      });

      const interval = 0;
      await syncWorker.intervalSync({ url: 'url' }, interval);
      expect(syncWorker.syncronize).toHaveBeenCalledWith({ url: 'url' });
      expect(syncWorker.syncronize).toHaveBeenCalledTimes(3);
    });

    it('should login when a sync response its "Unauthorized"', async () => {
      spyOn(syncWorker, 'login').and.returnValue(Promise.resolve());
      let syncCalls = 0;
      spyOn(syncWorker, 'syncronize').and.callFake(() => {
        if (syncCalls === 1) {
          syncWorker.stop();
        }
        syncCalls += 1;
        return Promise.reject({ status: 401, message: 'error' }); // eslint-disable-line prefer-promise-reject-errors
      });

      const interval = 0;
      await syncWorker.intervalSync({ url: 'url' }, interval);
      expect(syncWorker.login).toHaveBeenCalledWith('url', 'admin', 'admin');
    });
  });

  describe('login', () => {
    it('should login to the target api and set the cookie', async () => {
      backend.restore();
      backend.post('http://localhost/api/login', { body: '{}', headers: { 'set-cookie': 'cookie' } });
      spyOn(request, 'cookie');
      await syncWorker.login('http://localhost', 'username', 'password');
      expect(request.cookie).toHaveBeenCalledWith('cookie');
    });

    it('should catch errors and log them', async () => {
      spyOn(request, 'post').and.callFake(() => Promise.reject(new Error('post failed')));
      await syncWorker.login('http://localhost', 'username', 'password');
      expect(errorLog.error.calls.argsFor(0)[0]).toMatch('post failed');
    });
  });

  describe('start', () => {
    it('should not fail on sync not in settings', async () => {
      await settingsModel.db.update({}, { $unset: { sync: '' } });
      spyOn(syncWorker, 'intervalSync');
      const interval = 2000;

      let thrown;
      try {
        await syncWorker.start(interval);
      } catch (e) {
        thrown = e;
      }
      expect(thrown).not.toBeDefined();
    });
    it('should get sync config and start the sync', async () => {
      spyOn(syncWorker, 'intervalSync');
      const interval = 2000;

      await syncWorker.start(interval);
      expect(syncWorker.intervalSync).toHaveBeenCalledWith({ url: 'url', active: true, config: {} }, interval);
    });

    describe('when there is no sync config', () => {
      it('should not start the intervalSync', async () => {
        spyOn(syncWorker, 'intervalSync');
        await settings.save({ sync: {} });
        await syncWorker.start();
        expect(syncWorker.intervalSync).not.toHaveBeenCalled();
      });
    });
  });
});
