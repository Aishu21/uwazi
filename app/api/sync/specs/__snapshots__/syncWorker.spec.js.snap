// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syncWorker login should catch errors and log them 1`] = `
Array [
  "
Error: post failed
    at Object.spyOn.and.callFake (/Users/daneryl/workspace/uwazi/app/api/sync/specs/syncWorker.spec.js:153:64)
    at Object.SpyStrategy.exec (/Users/daneryl/workspace/uwazi/node_modules/jest-jasmine2/build/jasmine/spy_strategy.js:53:17)
    at Object.spy [as post] (/Users/daneryl/workspace/uwazi/node_modules/jest-jasmine2/build/jasmine/create_spy.js:67:42)
    at Object.post [as login] (/Users/daneryl/workspace/uwazi/app/api/sync/syncWorker.js:77:38)
    at Object.login (/Users/daneryl/workspace/uwazi/app/api/sync/specs/syncWorker.spec.js:154:24)
    at Object.asyncJestTest (/Users/daneryl/workspace/uwazi/node_modules/jest-jasmine2/build/jasmine_async.js:108:37)
    at resolve (/Users/daneryl/workspace/uwazi/node_modules/jest-jasmine2/build/queue_runner.js:56:12)
    at new Promise (<anonymous>)
    at mapper (/Users/daneryl/workspace/uwazi/node_modules/jest-jasmine2/build/queue_runner.js:43:19)
    at promise.then (/Users/daneryl/workspace/uwazi/node_modules/jest-jasmine2/build/queue_runner.js:87:41)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)",
]
`;
